docs/design/proposals/metro-dr.md

```diff
@@ -0,0 +1,70 @@
+corresponding PV name and the unique ID mapping, It creates the backend volume
+with the unique ID and returns the encoded value which contains all the
+required information for other operations. For metro DR, the same volume needs
+to be used in the secondary Kubernetes cluster. In the secondary cluster, the
+cephcsi cannot identify the volume corresponding to the PVC. Based on the
+assumption that the Kubernetes namespace and the PVC name remain the same when
+admin is doing the failover/fallback cephcsi can use the PVC name and namespace
+name to identify the RBD image. For that cephcsi needs to make a few changes.
+
+### Pre-req
+
+* The PVC name and namespace need to be passed to the CSI driver as part of the
+  volume create request.
+* A new storage-class parameter to indicate that image name need to be derived
+  from PVC name and namespace name.
+* An unique ID in the cephcsi ceph cluster configmap
+
+In the case of the primary cluster as the PVC name and the namespace name
+should be sent in volume create request, the cephcsi will generate omap data
+and creates its request name (PV name) and unique ID mapping in omap
+`csi.volume.<uniqueID>.PVName` : `volume UUID` (this is required to identify
+the RBD image/cephFS subvolume belongs to the PV) and creates an RBD image with
+the name `csi-vol-<pvcname>-<namespacename>-<UUID>`.  When doing the failover,
+cephcsi in the secondary cluster will check the volume is already present, If
+the image is present cephcsi will add its request name(PV Name) mapping to the
+UUID (`csi.volume.<uniqueID>.PVName` : `volume UUID`). As ID passed in
+configmap is unique across all the cephcsi instances accessing the ceph cluster
+will help cephcsi to identify the volume.
+
+Currently, we are making use of watchers in node stage request to make sure RWO
+PVC is mounted on a single node at a given point of time. We need to change the
+watchers logic in the node stage request as when we enable the RBD mirroring on
+an image a watcher will be added on an RBD image.
+
+In case of the PVC is Retained and recreated with the same name in the same
+namespace as the request name is unique cephcsi will not return the underlying
+storage. But if the PVC is Retained on one cluster and PVC gets created with
+the same name in the same namespace in another cluster will get underlying
+storage (it won't create a new volume (the support for the need to be added
+later by marking RBD/CephFS storage for not to be reused))
+
+### For snapshot
+
+This will be the same as the CreateVolume request, here instead of the
+storageclass parameter, cephcsi will check the snapshotclass parameter.
+
+Same as PVC name and namespace name to generate the volume name. We will use
+snapshot name and snapshot namespace name to generate the snapshot name, created
+an [issue](https://github.com/kubernetes-csi/external-snapshotter/issues/341)
+in external snapshotter to pass extra snapshot metadata to create the snapshot
+request.
```

Comment:
I would like to add a scenarios section here, with how this would be used. Something along the following lines,

```
**Setup:**
**Scenario 1: Single shared pool across peer kubernetes clusters:**
  - Kubernetes cluster-a
    - ceph-csi config map
      - clusterID: ceph-cluster-foo
        - uniqueID: bar
  - Kubernetes cluster-b
    - ceph-csi config map
      - clusterID: ceph-cluster-foo
        - uniqueID: baz
  - Kubernetes cluster-c
    - ceph-csi config map
      - clusterID: ceph-cluster-foo
        - uniqueID: qux
  - StorageClass-ceph:
    - clusterID: ceph-cluster-foo
    - PVCNaming: true
  PVC:
    - StorageClaassName: StorageClass-ceph

**Scenario 2: Affined pool (pool per peer kubernetes cluster):**
  - Kubernetes cluster-a
    - ceph-csi config map
      - clusterID: ceph-cluster-foo
        - uniqueID: bar
  - Kubernetes cluster-b
    - ceph-csi config map
      - clusterID: ceph-cluster-foo
        - uniqueID: baz
  - Kubernetes cluster-c
    - ceph-csi config map
      - clusterID: ceph-cluster-foo
        - uniqueID: qux
  - SC-a:
    - clusterID: ceph-cluster-foo
    - pool: pool-affined-a
    - PVCNaming: true
  - SC-b:
    - clusterID: ceph-cluster-foo
    - pool: pool-affined-b
    - PVCNaming: true
  - SC-c:
    - clusterID: ceph-cluster-foo
    - pool: pool-affined-c
    - PVCNaming: true
  PVC:
    - SC: SC-a/SC-b/SC-c
      - Possibly derived using kustomize during initial deployment to a chosen affined StorageClass for the cluster

**Scenario 3: Adding a peer cluster to an existing kubernetes cluster with existing provisioned PVCs**
  - NOTE: As existing ceph-csi would already have used the empty `uniqueID`, only one such cluster can be added as a peer to reuse existing provisioned PVCs. Adding more than one such peer would result in more than one of them using the empty `uniqueID` and will result in deployments not getting their storage bound when migrated to peer clusters.
  - Kubernetes cluster-a
    - ceph-csi config map
      - clusterID: ceph-cluster-foo
      - NOTE: no `uniqueID` as cluster was setup prior to DR needs
  - Kubernetes cluster-b
    - ceph-csi config map
      - clusterID: ceph-cluster-foo
        - uniqueID: baz
  - SC-a:
    - clusterID: ceph-cluster-foo
    - PVCNaming: true
      - NOTE: The above parameter should be set prior to provisioning volumes that may in the future be migrated to peer kubernetes clusters, as it is used as the identifying signature to find backend images for the request
  PVC:
    - SC: SC-a
```